### DISCUSSION TITLE
Avoiding the use of Any

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [python/typing#1246] [Q&A]
This is a function to validate incoming objects as lists of strings:
```
def validate_list(this_object: Any) -> Any:
    if isinstance(this_object, list):
        print('this_object is a list ... hooray')
        return this_object
    else:
        print('this_object is not a list ... grimbo :(')
        dodgy_object: Any = cast(Any, this_object)
        return dodgy_object
```
For brevity, the code is not checking if this_object contains strings.

If this_object is not a list then it is cast as Any to match the function's return type.

If this_object is a list then it is also returned as an Any object. 

If allowed, I could use Union[Any, list[str]] as the return type. If so, isn't Union[Any] the same as Union[Any, list[str]]?

Finally, how do you avoid the use of Any in this case? Thanks

**Comment by bschnurr** [comment]
sometimes to avoid union returns you can use overloads

maybe something like this

```python
from typing import Any, cast, overload


@overload
def validate_list(this_object: list) -> list: ...
def validate_list(this_object: Any) -> Any:
    if isinstance(this_object, list):
        print('this_object is a list ... hooray')
        return this_object
    else:
        print('this_object is not a list ... grimbo :(')
        dodgy_object: Any = cast(Any, this_object)
        return dodgy_object
```

**Comment by dineshbvadhia** [comment]
Will look into @overload.

But, the code is still using Any ?

**Comment by bschnurr** [comment]
true, but it will help typecheckers when accessing the return type if you are passing in a known list

**Comment by debonte** [comment]
How are you planning to use this validation function? Maybe a [`TypeGuard`](https://docs.python.org/3/library/typing.html?highlight=typeguard#typing.TypeGuard) would work better?

**Comment by dineshbvadhia** [comment]
I've built an api where the inputs are different types of objects including lists of strings. The validation functions ensure the input is correct. If possible, I'd like to avoid the use of Any.

The other reason to eliminate the use of Any is because the plan is to test the code base with mypyc. The mypyc documentation discourages the use of Any.

**Reply by erictraut** [reply]
What do you intend for the validation function to do at runtime when it detects that the object doesn't match the validation criteria? Should it 1) return False, 2) raise an exception, or 3) have some other side effect like logging, but otherwise simply return the object as is?

Depending on your intent, you might use one of the following:

```python
from typing import Any, TypeGuard, TypeVar

def is_list_of_strs(obj: object) -> TypeGuard[list[str]]:
    return isinstance(obj, list) and all([isinstance(x, str) for x in obj])

def validate_list(obj: object) -> list[Any]:
    if isinstance(obj, list):
        return obj
    raise TypeError("Not a list")

_T = TypeVar("_T")

def check_list(obj: _T) -> _T:
    if isinstance(obj, list):
        print("Is a list")
    else:
        print("Not a list")

    return obj
```

You should never need to `cast` to `Any` because all types are compatible with `Any`. That's what makes `Any` so dangerous because it effectively disables all type violation checks.

**Reply by dineshbvadhia** [reply]
Ooops, see response below

**Comment by dineshbvadhia** [comment]
Re: What do you intend for the validation function to do at runtime when it detects that the object doesn't match the validation criteria? 

2) Raise an exception. 
And, return a message to the api caller saying why.

Thank-you for the code and will try it out.
